Fix descriptions to sync with tooltip. Revamped content. (#81)
wso2/docs-choreo-dev: en/docs/apis/api-concepts.md

# API Concepts

> Choreo APIs allow you to create APIs, manage APIs, and expose services and external APIs as managed APIs. Here are some basic concepts which will help you understand Choreo APIs as you create and manage your APIs in the [Choreo Console]({{choreo_console}}).

Reviewer: "Choreo APIs " here does not sound right. Shall we say "Choreo’s API management capabilities allow . . ."

Author: +1

 
## API

> An API is an intermediate layer that acts as a communication protocol between a consumer and a service, simplifying the consumption of the service. In addition to hiding the underlying implementation details of a service, an API provides a secure, controlled, and well-documented approach for accessing the exposed service.

Reviewer: Shall we say "approach to access an exposed service" instead of "approach for accessing the exposed service".

Author: +1

> 

Reviewer 2: Don't we need to include concepts defined in https://apim.docs.wso2.com/en/latest/get-started/key-concepts/ ?
However some of them like API Product, API Visibility, Workflows, Message Mediation Policies, Handler, Tenant, Service Catalog and roles might not be relevant to choreo as of now.

Author: Please let us know the list that applies to Choreo, and then we can proceed to add it here.

Reviewer 2: Excluded list is mentioned above. So we can add the remaining items.

Author: Thanks. Will add the rest.

## API Management Components

> The Choreo API management is carried out by the key components: [API Publisher](#api-publisher), [Developer Portal](#developer-portal), [API Key Manager](#api-key-manager),and [API Gateway](#api-gateway).  

Reviewer: "The Choreo API management" does not sound right here.  Shall we say "The key components of Choreo API managements are . . ."

Author: Changed to "API management in Choreo is facilitated by the key components: [API Publisher](#api-publisher), [Developer Portal](#developer-portal), [API Key Manager](#api-key-manager),and [API Gateway](#api-gateway).  "

> API management in Choreo is facilitated by the key components: [API Publisher](#api-publisher), [Developer Portal](#developer-portal), [API Key Manager](#api-key-manager),and [API Gateway](#api-gateway).  

Reviewer 3: ```suggestion
API management in Choreo is facilitated by the key components: [API Publisher](#api-publisher), [Developer Portal](#developer-portal), [API Key Manager](#api-key-manager), and [API Gateway](#api-gateway).  
```

Author: fixed

> API management in Choreo is facilitated by the key components: [API Publisher](#api-publisher), [Developer Portal](#developer-portal), [API Key Manager](#api-key-manager),and [API Gateway](#api-gateway).  

Reviewer 2: We need to add Traffic Manager as well. Also it should have a sub section similar to what we have in APIM docs

Author: done.

### API Publisher 
The API Publisher is designed for API creators to develop, document, secure, test, and version APIs with ease. It also caters to more API management-related tasks, such as allowing API Publishers to deploy and publish APIs and applying rate-limiting policies.

### Developer Portal
The Developer portal allows API publishers to host and advertise their APIs and API consumers to discover, evaluate, subscribe to, and consume APIs securely.

### API Key Manager

> The API Key Manager is the identity provider for Choreo and acts as the Secure Token Service (STS). Choreo's API management supports OAuth2.0 and API-Key-based authentication mechanisms.

Reviewer 2: identity provider for Choreo **API Management**

Author: Fixed.

In Choreo, tokens are bound to an application. The Key manager provides a token API to generate access tokens. Clients can use these tokens to invoke APIs exposed by Choreo. The Key Manager also exposes a revoke token API that clients can use to revoke an access token. A client can generate an OAuth2.0 access token by invoking the token API directly or via the Developer Portal UI. Alternatively, a client can generate an API Key through the Developer Portal without calling the Key Manager. The API Key is a self-signed JWT token. When a client invokes an API with an OAuth2.0 access token or an API Key, the Gateway validates the token by validating its signature and subscription.

> The Key Manager performs scope validation as well.

Reviewer: Need to get a confirmation from the Choreo APIM team on accuracy of technical details.

Author: Ack

> The Key Manager performs scope validation as well.

Reviewer 2: `The Key Manager performs scope validation as well.`

Lets remove this since we do not support scope validation ATM

Author: done

### API Gateway
API Gateway acts as the entry point for an API request. It secures, protects, manages, and scales API calls. 

The API Gateway does the JWT token validation by validating the signature, issuer, expiry time, and subscription. The subscription is validated using the in-memory map. This in-memory map includes API-related information, application-related information, subscription-related information, etc., and is updated each time an artifact (API/application) is updated.

Once the token is validated, the API Gateway acts upon the API request before sending it to the backend. It first processes the message to a pre-configured format (e.g., JSON, XML, CSV, etc.). It then applies security policies, rate-limiting policies, collects statistics, etc., via its handlers. The mediators then act upon the API payload based on the mediation logic developed. Then, the message is formatted to a pre-configured format (e.g., JSON, XML, CSV, etc.) and sent to the backend. 

> 

Reviewer: Need to get a confirmation from the Choreo APIM team on accuracy of technical details.

Author: Ack

> Once the token is validated, the API Gateway acts upon the API request before sending it to the backend. It first processes the message to a pre-configured format (e.g., JSON, XML, CSV, etc.). It then applies security policies, rate-limiting policies, collects statistics, etc., via its handlers. The mediators then act upon the API payload based on the mediation logic developed. Then, the message is formatted to a pre-configured format (e.g., JSON, XML, CSV, etc.) and sent to the backend. 

Reviewer 2: Shall we verify this with @Rajith90. AFAIK in MGW (envoy) we don't do such conversion. Also we don't support any mediations in MGW and essentially in Choreo.

## API Lifecycle
API lifecycle comprises of the stages an API goes through from creation to retirement. APIs have lifecycle stages that are independent of their backend services. API lifecycle is independent of its backend service. The lifecycle states are `Created`, `Prototyped`, `Published`, `Deprecated`, `Blocked`, and `Retired`.

> 

Reviewer: need to remove the repetitive sentence on "API lifecycle is independent of its backend service."

Author: fixed

In Choreo, tokens are bound to an application. The Key manager provides a token API to generate access tokens. Clients can use these tokens to invoke APIs exposed by Choreo. The Key Manager also exposes a revoke token API that clients can use to revoke an access token. A client can generate an OAuth2.0 access token by invoking the token API directly or via the Developer Portal UI. Alternatively, a client can generate an API Key through the Developer Portal without calling the Key Manager. The API Key is a self-signed JWT token. When a client invokes an API with an OAuth2.0 access token or an API Key, the Gateway validates the token by validating its signature and subscription.

The Key Manager performs scope validation as well.

### Choreo Connect

Choreo Connect is an API gateway designed for microservices that is cloud-native, decentralized, and developer-centric. This API gateway is a lightweight message processor for APIs that facilitates message security, transport security, routing, and other API Management functions such as throttling, rate-limiting, etc.

Choreo Connect consists of three components: Router, Enforcer, and Adapter.

> ![Choreo Connect](../assets/img/apis/choreo-connect.png){:style="height:75%;width:90%"}

Reviewer 3: Shouldn't we re-draw this with choreo colors? Otherwise having this in orange won't suit for choreo docs.

Author: Holding this off, until we figure out the colors we want to use for Choreo

# API Concepts
Choreo's API management capabilities allow you to create APIs, manage APIs, and expose services and external APIs as managed APIs. Here are some basic concepts which will help you understand API management in Choreo as you create and manage your APIs in the [Choreo Console]({{choreo_console}}).
 
## API
An API is an intermediate layer that acts as a communication protocol between a consumer and a service, simplifying the consumption of the service. In addition to hiding the underlying implementation details of a service, an API provides a secure, controlled, and well-documented approach to access an exposed service.

## API format

> OpenAPI Specification (OAS, a.k.a Swagger) format is the underlying representation of an API in Choreo. You can import an API using a Swagger definition in the Choreo console. Additionally, a user can view, edit, import, or download an API definition in OAS format via the Choreo console.

Reviewer 3: Shouldn't we mention that users can create an API  from scratch too?

Author: You can't actually create an API from scratch in swagger. You have to create an API, and then view the source, right?

Reviewer 3: In choreo we allow creating from scratch as well. This is high level API concepts page. So why we are talking only about creating from OpenAPI stuff here. We need to reword mentioning the other option as well. 
<img width="1606" alt="image" src="https://user-images.githubusercontent.com/16300038/120242966-e72e2f00-c283-11eb-9ac9-b65f73512f45.png">

Author: You have to design the API from scratch via the UI. It doesn't open you the swagger editor to edit the API. Here we are talking about what you can do with the API definition.  And we can't create an API from scratch through the editor, by writing the API definition. The underlying format is the OpenAPI format.

Reviewer 3: Yes. Was thinking since this is highlevel API concepts page, shouldn't we at least mention about some concepts related to the ability to start from scratch (not under this sub-topic) . But if we don't have anything to say specifically, lets go with the current one.

Author: We are mentioning it in the first sentence in the document. "Choreo's API management capabilities allow you to create APIs, manage APIs, and expose services and external APIs as managed APIs. " So I guess we are good to go.

## API resource path and HTTP methods

> An API comprises one or more resources, each of which has a unique resource path (URI). An API Resource has a set of HTTP methods that operates on it. The supported HTTP methods are `GET`, `POST`, `PUT`, `DELETE`, `PATCH`, `HEAD`, and `OPTIONS`.

Reviewer 3: Since we use `URI` within brackets shouldn't we use the exact name i.e. `Unique Resource Identifire` instead of `unique resource path`

Author: On second thoughts, spelling out URI is not actually needed. https://developers.google.com/style/abbreviations

## API lifecycle
An API lifecycle comprises the stages an API goes through from creation to retirement. API lifecycle is independent of its backend service. The lifecycle states are: `Created`, `Prototyped`, `Published`, `Deprecated`, `Blocked`, `Deployed`, and `Retired`.

## Application
An Application in Choreo is the logical representation of a physical application such as a mobile app, web app, device, etc. For an application to use an API, the application should subscribe to the relevant APIs it intends to use. A subscription to an API happens over a selected business plan.  The business plan determines the usage quota the subscription gets. An application has a consumer key and a consumer secret which acts as the credentials of the application.

## Access token
Choreo uses the OAuth2.0 bearer token-based authentication mechanism to allow a consumer to access an API.  
The access token is a string that the consumer passes in an HTTP header of the API request. Choreo supports JWT formatted self-contained access tokens.

## Rate limiting
Rate limiting makes it possible to limit the number of permitted requests to an API for a given time window.

> Thereby it allows you to:

Reviewer 3: `When a colon introduces a list, the text that precedes the colon must be able to stand alone as a complete sentence.`
Seems we need to follow the above for this sentence

Author: reworded.

# API Concepts
Choreo's API management capabilities allow you to create APIs, manage APIs, and expose services and external APIs as managed APIs. Here are some basic concepts which will help you understand API management in Choreo as you create and manage your APIs in the [Choreo Console]({{choreo_console}}).
 
## API
An API is an intermediate layer that acts as a communication protocol between a consumer and a service, simplifying the consumption of the service. In addition to hiding the underlying implementation details of a service, an API provides a secure, controlled, and well-documented approach to access an exposed service.

> ## API Format

Reviewer: Since we are going with title-style capitalization for the page title and sentence-style capitalization for all other subsection headings, might have to follow the same for all subsection headings on this page.

Author: done.

> Open API Specification (OAS, a.k.a Swagger) format is the underlying representation of an API in Choreo. You can import an API using a Swagger definition in the Publisher Portal. Additionally, a user can view, edit, import, or download an API definition in OAS format via the API Publisher Portal.

Reviewer: Should we say OpenAPI instead of Open API?

Author: Yes. Fixed.

## API Resource path and HTTP Methods

> An API comprises one or more resources, each of which has a unique resource path (URI). An API Resource has a set of HTTP methods that operates on it. The supported HTTP methods are GET, POST, PUT, DELETE, PATCH, HEAD, and OPTIONS.

Reviewer: Should we format the HTTP methods using backticks?

Author: Shouldn't it be bold? https://docs.microsoft.com/en-us/style-guide/developer-content/formatting-developer-text-elements

Author: added back ticks.

## API Lifecycle
API lifecycle comprises of the stages an API goes through from creation to retirement. API lifecycle is independent of its backend service. The lifecycle states are `Created`, `Prototyped`, `Published`, `Deprecated`, `Blocked`, `Deployed`, and `Retired`.

## Application
An Application in Choreo is the logical representation of a physical application such as a mobile app, web app, device, etc. For an application to use an API, the application should subscribe to the relevant APIs it intends to use. A subscription to an API happens over a selected business plan.  The business plan determines the usage quota the application gets. An application has a consumer key and a consumer secret which acts as the credentials of the application.

## Access Token
Choreo uses the OAuth2.0 bearer token-based authentication mechanism to allow a consumer to access an API.  
The access token is a string that the consumer passes in an HTTP header of the API request. Choreo supports JWT formatted self-contained access tokens.

## Rate Limiting

> Rate limiting allows you to limit the number of permitted requests to an API for a given time window. Rate limiting provides you to:

Reviewer: Doesn't seem correct to say "Rate limiting provides you to:"
We might have to go with "Rate limiting allows you to:" here and perhaps consider changing the previous sentence as "Rate limiting makes it possible to limit the number of . . . ."

Author: done

# API Concepts
Choreo's API management capabilities allow you to create APIs, manage APIs, and expose services and external APIs as managed APIs. Here are some basic concepts which will help you understand API management in Choreo as you create and manage your APIs in the [Choreo Console]({{choreo_console}}).
 
## API
An API is an intermediate layer that acts as a communication protocol between a consumer and a service, simplifying the consumption of the service. In addition to hiding the underlying implementation details of a service, an API provides a secure, controlled, and well-documented approach to access an exposed service.

## API format

> OpenAPI Specification (OAS, a.k.a Swagger) format is the underlying representation of an API in Choreo. You can import an API using a Swagger definition in the Publisher Portal. Additionally, a user can view, edit, import, or download an API definition in OAS format via the API Publisher Portal.

Reviewer 4: ```suggestion
OpenAPI Specification (OAS, a.k.a Swagger) format is the underlying representation of an API in Choreo. You can import an API using a Swagger definition in the Choreo console. Additionally, a user can view, edit, import, or download an API definition in OAS format via the API Publisher Portal.
```

Author: applied change

## API resource path and HTTP methods
An API comprises one or more resources, each of which has a unique resource path (URI). An API Resource has a set of HTTP methods that operates on it. The supported HTTP methods are GET, POST, PUT, DELETE, PATCH, HEAD, and OPTIONS.

## API lifecycle
An API lifecycle comprises the stages an API goes through from creation to retirement. API lifecycle is independent of its backend service. The lifecycle states are: Created, Prototyped, Published, Deprecated, Blocked, Deployed, and Retired.

## Application

> An Application in Choreo is the logical representation of a physical application such as a mobile app, web app, device, etc. For an application to use an API, the application should subscribe to the relevant APIs it intends to use. A subscription to an API happens over a selected business plan.  The business plan determines the usage quota the application gets. An application has a consumer key and a consumer secret which acts as the credentials of the application.

Reviewer 4: ```suggestion
An Application in Choreo is the logical representation of a physical application such as a mobile app, web app, device, etc. For an application to use an API, the application should subscribe to the relevant APIs it intends to use. A subscription to an API happens over a selected business plan.  The business plan determines the usage quota the subscription gets. An application has a consumer key and a consumer secret which acts as the credentials of the application.
```

Author: applied change

The access token is a string that the consumer passes in an HTTP header of the API request. Choreo supports JWT formatted self-contained access tokens.

## Rate limiting
Rate limiting makes it possible to limit the number of permitted requests to an API for a given time window.
Thereby it allows you to:

   - Protect your APIs from common types of security attacks, such as certain types of Denial of Service (DoS) attacks.
   - Regulate traffic according to infrastructure availability.
   
## Tags
Tags allow API providers to categorize APIs that have similar attributes. When a tagged API gets published to the API Developer Portal, its tags appear as clickable links. API consumers can use the link to navigate to a category of interest. API consumers can also search APIs that match a particular tag on the Developer Portal.

## API management components
API management in Choreo is facilitated by the key components: [API Publisher](#api-publisher), [Developer Portal](#developer-portal), [API Key Manager](#api-key-manager), [Choreo Connect](#choreo-connect), and [Traffic Manager](#traffic-manager).  

> ### API publisher 

Reviewer 4: This should be Choreo Console

### Choreo connect

Choreo Connect is an API gateway designed for microservices that is cloud-native, decentralized, and developer-centric. This API gateway is a lightweight message processor for APIs that facilitates message security, transport security, routing, and other API Management-related quality of services such as throttling, rate-limiting, etc.

Choreo Connect consists of three components: Router, Enforcer, and Adapter.

![Choreo Connect](../assets/img/apis/choreo-connect.png){:style="height:75%;width:90%"}

#### Router

The Router in Choreo Connect uses the [Envoy Proxy](https://www.envoyproxy.io/) and is primarily responsible for routing traffic to the relevant endpoints. The Router is also capable of exposing the APIs in Choreo to make them discoverable.

#### Enforcer

> The Enforcer is responsible for carrying out API management-related quality-of-services such as controlling access, enabling security, publishing analytics, etc. When the Router receives a request, it sends it to the Enforcer. The Enforcer performs the relevant API management-related quality of services on the request and replies to the Router. Based on the Router's response, the Enforcer will decide to allow or deny the request to pass to the backend.

Reviewer 4: ```suggestion
The Enforcer is responsible for carrying out API management-related quality-of-services such as controlling access, enabling security, publishing analytics, etc. When the Router receives a request, it sends it to the Enforcer. The Enforcer performs the relevant API management-related quality of services on the request and replies to the Router. Based on the Enforcer's response, the Router will decide to allow or deny the request to pass to the backend.
```

Author: changed

#### Adapter 

> The Adapter is responsible for converting the API definition to the respective format understood by the Router and the Enforcer. The Adapter will pass the data to and from the Router and Enforcer.

Reviewer 4: ```suggestion
The Adapter is responsible for converting the API definition to the respective format understood by the Router and the Enforcer. Adapter gets notified when changes happens in the control plane, and it passes these changes to Enforcers and Routers connected to it via the XDS protocol
```

Author: changed

### Traffic manager 

The Traffic Manager helps regulate API traffic, makes APIs and applications available to consumers at different service levels, and secures APIs against security attacks. The Traffic Manager features a dynamic throttling engine that processes rate-limiting policies in real-time. 

> Additionally,  the Traffic Manager keeps the API Gateway's in-memory map, which it uses for key validation, up-to-date via a JMS topic. The Traffic Manager publishes artifact (API/application) update events that it receives from the API Publisher and API Developer Portal to a JMS topic. The API Gateway receives these events via the JMS topic and updates its in-memory map.

Reviewer 4: Shall we add this content inside the APIM control plane, rather than adding this inside traffic manager.  Because TM will run in data plane, the in memory data are stored in the control plane component.

Author: Removed the second paragraph, as that might confuse a user regarding the architecture. We will discuss this in the architecture section when we include it.
